src/entities.py

```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env python
+from __future__ import print_function
+import json
+import re
+
+### Declare headers
+header_a = """/*
+ * Generated file - do not edit directly.
+ * This file was generated from:
+ *     https://html.spec.whatwg.org/entities.json
+ * by means of entities.py in the libgd src dir
+*/
+"""
+
+header_b = """
+#include "entities.h"
+
+const struct entities_s entities[NR_OF_ENTITIES] = {"""
+
+header_c = """/*
+ * Generated file - do not edit directly.
+ * This file was generated from:
+ *     https://html.spec.whatwg.org/entities.json
+ * by means of entities.py in the libgd src dir
+*/
+#ifndef GD_ENTITIES_H
+
+#include <stdlib.h>
+#include <stdint.h>
+
+//html entity strings are entity prefix + string + suffix limited
+//hex and dec should be limited to current unicode spec + entity prefix + suffix
+"""
+
+header_d = """
+struct entities_s{
+                  	char *name;
+                        uint32_t unicode_a;
+                        uint32_t unicode_b;
+                };
+extern const struct entities_s entities[NR_OF_ENTITIES];
+
+#endif
+"""
+
+### Open files
+file_json = open("entities.json", mode="r")
+file_ent_h = open("entities.h", mode="w")
+file_ent_c = open("entities.c", mode="w")
+
+### Initialize some objects
+x = json.load(file_json)
+curline = 1
+totallines = 0
+len_name_max = 0
+p = re.compile("&\S+;")
+
+### Print file headers
+file_ent_c.write(header_a + header_b + "\n")
+file_ent_h.write(header_c)
+file_ent_h.write("#define GD_ENTITIES_H 1\n")
+
+### Count how many matching lines we have
+for e in x:
+    if p.match(e):
+        totallines+=1
+### Print json entities to struct
+for e in x:
+    if p.match(e):
+        if len(e) > len_name_max:
+            len_name_max = len(e)
+        string = "\t\t\t" + "{\"" + e.replace("&", "") + "\", "
+        file_ent_c.write(string.replace(";", ""))
+        if len(x[e]["codepoints"]) > 1:
+            string = str(x[e]["codepoints"][0]) + ", " + str(x[e]["codepoints"][1]) + "}"
```

Comment:
this for loop could be simplified with a little data manipulation

```
codepoints = x[e]["codepoints"]
# If there's only one codepoint, make sure the 2nd one is zero.  Otherwise this is ignored.
codepoints.append(0)
file_ent_c.write(str(codepoints[0]) + ", " + str(codepoints[1]) + "},\n")
```